master/internal/api_generic_tasks.go

```diff
@@ -0,0 +1,791 @@
+	}
+
+	if err := a.canCreateGenericTask(ctx, projectID); err != nil {
+		return nil, err
+	}
+
+	// forkedConfig denotes the config of the task we are forking from
+	var forkedConfig []byte
+	var forkedContextDirectory []byte
+	if req.ForkedFrom != nil {
+		getTaskReq := &apiv1.GetGenericTaskConfigRequest{
+			TaskId: *req.ForkedFrom,
+		}
+		resp, err := a.GetGenericTaskConfig(ctx, getTaskReq)
+		if err != nil {
+			return nil, err
+		}
+
+		forkedConfig = []byte(resp.Config)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(req.ContextDirectory) == 0 {
+			contextDirectoryResp, err := a.GetTaskContextDirectory(ctx, &apiv1.GetTaskContextDirectoryRequest{
+				TaskId: *req.ForkedFrom,
+			})
+			if err != nil {
+				return nil, err
+			}
+			forkedContextDirectory = []byte(contextDirectoryResp.B64Tgz)
+		}
+	}
+
+	if len(forkedConfig) == 0 && len(req.Config) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "No config file nor forked task provided")
+	}
+	configBytes, err := getConfigBytes([]byte(req.Config), forkedConfig)
+	if err != nil {
+		return nil, err
+	}
+	genericTaskSpec, warnings, contextDirectoryBytes, err := a.getGenericTaskLaunchParameters(
+		ctx, req.ContextDirectory, projectID, configBytes,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if req.InheritContext != nil && *req.InheritContext {
+		if req.ParentId == nil {
+			return nil, fmt.Errorf("could not inherit config directory since no parent task id provided")
+		}
+		contextDirectoryBytes, err = db.NonExperimentTasksContextDirectory(ctx, model.TaskID(*req.ParentId))
+		if err != nil {
+			return nil, err
+		}
+	}
+	if len(contextDirectoryBytes) == 0 {
+		contextDirectoryBytes = forkedContextDirectory
+	}
+
+	if err := check.Validate(genericTaskSpec.GenericTaskConfig); err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"invalid generic task config: %s",
+			err.Error(),
+		)
+	}
+
+	// Persist the task.
+	taskID := model.NewTaskID()
+	jobID := model.NewJobID()
+	startTime := time.Now()
+	err = db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		if err := db.AddJobTx(ctx, tx, &model.Job{
+			JobID:   jobID,
+			JobType: model.JobTypeGeneric,
+			OwnerID: &genericTaskSpec.Base.Owner.ID,
+		}); err != nil {
+			return fmt.Errorf("persisting job %v: %w", taskID, err)
+		}
+
+		configBytesJSON, err := yaml.YAMLToJSON(configBytes)
+		if err != nil {
+			return err
+		}
+		if err := db.AddTaskTx(ctx, tx, &model.Task{
+			TaskID:     taskID,
+			TaskType:   model.TaskTypeGeneric,
+			StartTime:  startTime,
+			JobID:      &jobID,
+			LogVersion: model.CurrentTaskLogVersion,
+			ForkedFrom: req.ForkedFrom,
+			Config:     ptrs.Ptr(string(configBytesJSON)),
+			ParentID:   (*model.TaskID)(req.ParentId),
+			State:      ptrs.Ptr(model.TaskStateActive),
+			NoPause:    req.NoPause,
+		}); err != nil {
+			return fmt.Errorf("persisting task %v: %w", taskID, err)
+		}
+
+		// TODO persist config elemnts
```

Comment:
what does this mean?

Reply:
Old todo that was still there for storing the config of the task. Fixed